### PR TITLE
Merge Map and Set Guards with Container

### PR DESCRIPTION
**Pull request by Achie72**
Remove JERRY_BUILTIN_MAP/SET/WEAKMAP/WEAKSET
and replace them with JERRY_BUILTIN_CONTAINER.

JerryScript-DCO-1.0-Signed-off-by: Bela Toth tbela@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/config.h [hunk 1]
```diff
@@ -602,11 +574,7 @@
 #endif
 #if (JERRY_ESNEXT == 0) \
 && ((JERRY_BUILTIN_DATAVIEW == 1) \
-|| (JERRY_BUILTIN_MAP == 1) \
-|| (JERRY_BUILTIN_SET == 1) \
-|| (JERRY_BUILTIN_WEAKMAP == 1) \
-|| (JERRY_BUILTIN_WEAKREF == 1) \
```
Comment (zherczeg) [inline jerry-core/config.h:608]: Don't remove weakref

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.inc.h [hunk 2]
```diff
@@ -220,40 +220,39 @@ OBJECT_VALUE (LIT_MAGIC_STRING_PROMISE_UL,
...
 
-#if JERRY_BUILTIN_WEAKMAP
 /* ECMA-262 v6, 23.1.1.1 */
 OBJECT_VALUE (LIT_MAGIC_STRING_WEAKMAP_UL,
               ECMA_BUILTIN_ID_WEAKMAP,
               ECMA_PROPERTY_CONFIGURABLE_WRITABLE)
-#endif /* JERRY_BUILTIN_WEAKMAP */
+
+/* ECMA-262 v6, 23.1.1.1 */
+OBJECT_VALUE (LIT_MAGIC_STRING_WEAKSET_UL,
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.inc.h:245]: This code added twice. Is this correct?

File: jerry-core/ecma/operations/ecma-objects.c [hunk 3]
```diff
@@ -2796,27 +2796,23 @@ ecma_object_check_class_name_is_object (ecma_object_t *obj_p) /**< object */
...
-#if JERRY_BUILTIN_MAP
+#if JERRY_BUILTIN_CONTAINER
           || ecma_builtin_is (obj_p, ECMA_BUILTIN_ID_MAP_PROTOTYPE)
 #if JERRY_ESNEXT
           || ecma_builtin_is (obj_p, ECMA_BUILTIN_ID_MAP_ITERATOR_PROTOTYPE)
 #endif /* JERRY_ESNEXT */
-#endif /* JERRY_BUILTIN_MAP */
-#if JERRY_BUILTIN_SET
+#endif /* JERRY_BUILTIN_CONTAINER */
+#if JERRY_BUILTIN_CONTAINER
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-objects.c:2805]: It seems these could be merged together (by removing this two line).

File: jerry-core/ecma/operations/ecma-objects.c [hunk 4]
```diff
@@ -2796,27 +2796,23 @@ ecma_object_check_class_name_is_object (ecma_object_t *obj_p) /**< object */
...
           || ecma_builtin_is (obj_p, ECMA_BUILTIN_ID_SET_PROTOTYPE)
 #if JERRY_ESNEXT
           || ecma_builtin_is (obj_p, ECMA_BUILTIN_ID_SET_ITERATOR_PROTOTYPE)
 #endif /* JERRY_ESNEXT */
-#endif /* JERRY_BUILTIN_SET */
-#if JERRY_BUILTIN_WEAKMAP
+#endif /* JERRY_BUILTIN_CONTAINER */
+#if JERRY_BUILTIN_CONTAINER
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-objects.c:2811]: Ditto.

**Review by zherczeg — CHANGES_REQUESTED** [review]
Good patch, only some minor changes.

File: jerry-core/api/jerry.c [hunk 5]
```diff
@@ -1820,21 +1818,13 @@ jerry_is_feature_enabled (const jerry_feature_t feature) /**< feature to check *
...
+#if JERRY_BUILTIN_CONTAINER
           || feature == JERRY_FEATURE_MAP
-#endif /* JERRY_BUILTIN_MAP */
-#if JERRY_BUILTIN_SET
           || feature == JERRY_FEATURE_SET
-#endif /* JERRY_BUILTIN_SET */
-#if JERRY_BUILTIN_WEAKMAP
           || feature == JERRY_FEATURE_WEAKMAP
-#endif /* JERRY_BUILTIN_WEAKMAP */
-#if JERRY_BUILTIN_WEAKREF
```
Comment (zherczeg) [inline jerry-core/api/jerry.c:1832]: Don't remove this.

File: jerry-core/ecma/base/ecma-gc.c [hunk 6]
```diff
@@ -1567,7 +1555,7 @@ ecma_gc_free_property (ecma_object_t *object_p, /**< object */
       break;
     }
 #endif /* JERRY_ESNEXT */
-#if JERRY_BUILTIN_WEAKMAP || JERRY_BUILTIN_WEAKSET || JERRY_BUILTIN_WEAKREF
+#if JERRY_BUILTIN_CONTAINER
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-gc.c:1558]: Keep `|| JERRY_BUILTIN_WEAKREF`

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: jerry-core/config.h [hunk 7]
```diff
@@ -111,30 +107,22 @@
...
 
-#ifndef JERRY_BUILTIN_WEAKSET
-# define JERRY_BUILTIN_WEAKSET JERRY_ESNEXT
-#endif /* !defined (JERRY_BUILTIN_WEAKSET) */
-
 #ifndef JERRY_MODULE_SYSTEM
 # define JERRY_MODULE_SYSTEM JERRY_ESNEXT
 #endif /* !defined (JERRY_MODULE_SYSTEM) */
 
+#ifndef JERRY_BUILTIN_CONTAINER
```
Comment (zherczeg) [inline jerry-core/config.h:122]: Maybe this could be moved up to its lexicographic place.

**Review by zherczeg — APPROVED** [review]
LGTM with minor change

**Review by rerobika — APPROVED** [review]

**Pull request merged**